cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```diff
@@ -230,10 +241,23 @@ public void launch(RuntimeJobInfo runtimeJobInfo) throws Exception {
       for (LocalFile fileToUpload : localFiles) {
         boolean isCacheable = !disableGCSCaching && fileToUpload instanceof CacheableLocalFile;
         String targetFilePath = getPath(isCacheable ? cacheRootPath : runRootPath, fileToUpload.getName());
-        uploadFutures.add(
-          provisionerContext.execute(() -> isCacheable ? uploadCacheableFile(bucket, targetFilePath, fileToUpload) :
-              uploadFile(bucket, targetFilePath, fileToUpload, false))
-            .toCompletableFuture());
+        String targetFilePathWithVersion = getPath(cacheRootPath, cdapVersion[0], fileToUpload.getName());
+
+        if (artifactsCacheablePerCDAPVersion.contains(fileToUpload.getName()) && !disableGCSCaching) {
+          // upload artifacts cacheable per cdap version to <bucket>/cdap-job/cached-artifacts/<cdapVersion>/
+          uploadFutures.add(
+            provisionerContext.execute(
+              () -> uploadCacheableFile(bucket, targetFilePathWithVersion, fileToUpload, cdapVersion))
+              .toCompletableFuture());
+        } else {
+          // upload cacheable artifacts to <bucket>/cdap-job/cached-artifacts/ and
+          // non-cacheable artifacts to <bucket>/cdap-job/<runid>/
+          uploadFutures.add(
+            provisionerContext.execute(() -> isCacheable ? uploadCacheableFile(bucket, targetFilePath,
```

Comment:
I probably would pull the `isCachable` outside to make the logic easier to follow:

```
if (cacheable) {
  uploadFutures.add(provisionerContext.execute(() -> uploadCacheableFile(...)));
} else {
  uploadFutures.add(provisionerContext.execute(() -> uploadFile(...)));
}
```

Also, variable name should be just `boolean cacheable` instead of `isCacheable`.

Reply:
ack